metadeploy/api/salesforce.py

```diff
@@ -0,0 +1,260 @@
+    features = scratch_org_definition.get("features", [])
+    create_args = {
+        "AdminEmail": email,
+        "ConnectedAppConsumerKey": SF_CLIENT_ID,
+        "ConnectedAppCallbackUrl": SF_CALLBACK_URL,
+        "Description": f"{repo_owner}/{repo_name} {repo_branch}",
+        # Override whatever is in scratch_org_config.days:
+        "DurationDays": DURATION_DAYS,
+        "Edition": scratch_org_definition["edition"],
+        "Features": ";".join(features) if isinstance(features, list) else features,
+        "HasSampleData": scratch_org_definition.get("hasSampleData", False),
+        "Namespace": (
+            cci.project_config.project__package__namespace
+            if scratch_org_config.namespaced
+            else None
+        ),
+        "OrgName": scratch_org_definition.get("orgName", "Metecho Task Org"),
+        # should really flesh this out to pass the other
+        # optional fields from the scratch org definition file,
+        # but this will work for a start
+    }
+    if SFDX_SIGNUP_INSTANCE:  # pragma: nocover
+        create_args["Instance"] = SFDX_SIGNUP_INSTANCE
+    response = devhub_api.ScratchOrgInfo.create(create_args)
+
+    # Get details and update scratch org config
+    return devhub_api.ScratchOrgInfo.get(response["id"])
+
+
+def _mutate_scratch_org(*, scratch_org_config, org_result, email):
+    """Set the scratch org config into a good state.
+
+    Update the org config for a new scratch org with details from its
+    ScratchOrgInfo.
+    """
+    scratch_org_config._scratch_info = {
+        "instance_url": org_result["LoginUrl"],
+        "org_id": org_result["ScratchOrg"],
+        "username": org_result["SignupUsername"],
+    }
+    scratch_org_config.config.update(scratch_org_config._scratch_info)
+    scratch_org_config.config.update(
+        {
+            "days": DURATION_DAYS,
+            "date_created": datetime.now(),
+            "created": True,
+            "email": email,
+            "scratch": True,
+            "is_sandbox": True,
+        }
+    )
+
+
+def _get_access_token(*, org_result, scratch_org_config):
+    """Trade the ScratchOrgInfo AuthCode for an org access token.
+
+    The AuthCode is short-lived so this is only useful immediately after
+    the scratch org is created. This must be completed once in order for
+    future access tokens to be obtained using the JWT token flow.
+    """
+    oauth = SalesforceOAuth2(
+        SF_CLIENT_ID, SF_CLIENT_SECRET, SF_CALLBACK_URL, scratch_org_config.instance_url
+    )
+    auth_result = oauth.get_token(org_result["AuthCode"]).json()
+    scratch_org_config.config["access_token"] = scratch_org_config._scratch_info[
+        "access_token"
+    ] = auth_result["access_token"]
+    scratch_org_config.config["refresh_token"] = scratch_org_config._scratch_info[
+        "refresh_token"
+    ] = auth_result["refresh_token"]
```

Comment:
```suggestion
    scratch_org_config.config["refresh_token"] = auth_result["refresh_token"]
```

The refresh_token isn't expected to be in ScratchOrgConfig._scratch_info.